src/app/core/device.ts

```diff
@@ -0,0 +1,1044 @@
+        this.__valveMap = valvemap;
+        this.__valveIs3DMap = isvalve3Ddata;
+    }
+
+    /**
+     * Returns the name of the device
+     *
+     * @returns {String}
+     * @memberof Device
+     */
+    get name(): string | String {
+        return this.__name.value;
+    }
+
+    /**
+     * Returns the list of layers in the device
+     * @return {Array}
+     * @memberof Device
+     */
+    get layers(): Array<Layer> {
+        return this.__layers;
+    }
+
+    /**
+     * Adds a connection to the device
+     * @param {Connection} connection
+     * @memberof Device
+     * @returns {void}
+     */
+    addConnection(connection: Connection): void {
+        this.__connections.push(connection);
+    }
+
+    /**
+     * Removes a connection from the device
+     * @param {Connection} connection
+     * @memberof Device
+     * @returns {void}
+     */
+    removeConnection(connection: Connection): void {
+        let i = this.__connections.indexOf(connection);
+        if (i != -1) {
+            this.__connections.splice(i, 1);
+        }
+    }
+
+    /**
+     * Adds a component to the device
+     * @param {Component} component Component to be added to the device
+     * @memberof Device
+     * @returns {void}
+     */
+    addComponent(component: Component): void {
+        if (component instanceof Component) {
+            this.__components.push(component);
+        } else {
+            throw new Error("Tried to add a component that isn't a component to the device");
+        }
+    }
+
+    /**
+     * Removes a component from the device
+     * @param {Component} component Component to remove from the device
+     * @memberof Device
+     * @returns {void}
+     */
+    removeComponent(component: Component): void {
+        //Remove the component from the map
+        let trydelete;
+        let componentid = component.getID();
+        let connectiontorefresh = null;
+
+        //Remove component from connections
+        for (let i in this.__connections) {
+            let connection = this.__connections[i];
+            try {
+                trydelete = connection.tryDeleteConnectionTarget(componentid);
+                if (trydelete) {
+                    console.log("Removed Component from Connection : ", connection.getID());
+                }
+            } catch (e) {
+                console.error(e);
+            }
+        }
+        //Check if the valve map has the component
+        let checkConnection: string | undefined = this.__valveMap.get(componentid);
+        if (checkConnection != undefined) {
+            connectiontorefresh = this.getConnectionByID(checkConnection);
+            this.__valveMap.delete(componentid);
+        }
+
+        let i = this.__components.indexOf(component);
+        if (i != -1) {
+            this.__components.splice(i, 1);
+        }
+
+        if (connectiontorefresh) {
+            Registry.viewManager.updatesConnectionRender(connectiontorefresh);
```

Comment:
Results in error "Object is possibly 'null'" because viewManager initialized to null in Registry's constructor. Unsure how to resolve